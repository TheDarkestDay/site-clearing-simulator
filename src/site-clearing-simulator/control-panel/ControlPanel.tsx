import { Card, Fab } from '@material-ui/core';
import { NavigationOutlined, RedoOutlined, SettingsBackupRestore, Stop, UndoOutlined } from '@material-ui/icons';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { isSimulationStopped } from '../store/selectors';
import { moveForward, rotateRight, rotateLeft, stopSimulation, resetSimulation } from '../store/store-slice';
import styles from './styles.module.css';

type Props = {
  className: string;
}

export const ControlPanel = ({ className }: Props) => {
  const dispatch = useDispatch();
  const isDisabled = useSelector(isSimulationStopped);

  const handleMoveForwardClick = () => {
    dispatch(moveForward());
  };

  const handleRotateLeftClick = () => {
    dispatch(rotateLeft());
  };

  const handleRotateRightClick = () => {
    dispatch(rotateRight());
  };

  const handleStopButtonClick = () => {
    dispatch(stopSimulation());
  };

  const handleStartNewSimulationButtonClick = () => {
    dispatch(resetSimulation());
  };

  return (
    <Card className={`${styles.controlPanel} ${className}`}>
      <div className={styles.controlButton}>
        <Fab color="primary" onClick={handleRotateLeftClick} disabled={isDisabled}>
          <UndoOutlined />
        </Fab>
      </div>

      <div className={styles.controlButton}>
        <Fab color="primary" onClick={handleMoveForwardClick} disabled={isDisabled}>
          <NavigationOutlined />
        </Fab>
      </div>

      <div className={styles.controlButton}>
        <Fab color="primary" onClick={handleRotateRightClick} disabled={isDisabled}>
          <RedoOutlined />
        </Fab>
      </div>

      <div className={styles.controlButton} onClick={handleStopButtonClick}>
        <Fab color="secondary" disabled={isDisabled}>
          <Stop />
        </Fab>
      </div>

      <div className={styles.controlButton} onClick={handleStartNewSimulationButtonClick}>
        <Fab color="secondary">
          <SettingsBackupRestore />
        </Fab>
      </div>
    </Card>
  );
};