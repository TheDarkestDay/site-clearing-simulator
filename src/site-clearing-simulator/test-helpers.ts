import { SiteClearingSimulatorState, BulldozerDirection } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};