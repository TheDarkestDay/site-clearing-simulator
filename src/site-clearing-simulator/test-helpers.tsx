import { createStore } from '@reduxjs/toolkit';
import { create } from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { GlobalState, rootReducer } from '../store';
import { SiteClearingSimulatorState, BulldozerDirection, initialState } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};

export const renderWithStore = (component: React.ReactNode, state: Partial<GlobalState['siteClearingSimulator']> = {}) => {
  const testStore = createStore(rootReducer, {
    siteClearingSimulator: {
      ...initialState,
      ...state,
    }
  });
  const consumer = (
    <Provider store={testStore}>
      {component}
    </Provider>
  );

  return create(consumer);
}