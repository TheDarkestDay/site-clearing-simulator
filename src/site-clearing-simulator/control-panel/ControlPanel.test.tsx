import { renderWithStore } from '../test-helpers';
import { ControlPanel } from './ControlPanel';

describe('ControlPanel component', () => {
  it('should match snapshot for ongoing simulation', () => {
    const wrapper = renderWithStore(<ControlPanel />, {
      bulldozerPosition: {
        x: 1,
        y: 1,
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot for stopped simulation', () => {
    const wrapper = renderWithStore(<ControlPanel />, {
      isStopped: true,
      bulldozerPosition: {
        x: 1,
        y: 1,
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot for bulldozer being on the starting position', () => {
    const wrapper = renderWithStore(<ControlPanel />);

    expect(wrapper).toMatchSnapshot();
  });
});