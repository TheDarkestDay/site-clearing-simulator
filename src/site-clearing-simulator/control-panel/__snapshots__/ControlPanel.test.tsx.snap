// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel component should match snapshot for bulldozer being on the starting position 1`] = `
<div
  className="MuiPaper-root MuiCard-root controlPanel undefined MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root Mui-disabled MuiFab-primary Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-primary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 7.27l4.28 10.43-3.47-1.53-.81-.36-.81.36-3.47 1.53L12 7.27M12 2L4.5 20.29l.71.71L12 18l6.79 3 .71-.71L12 2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root Mui-disabled MuiFab-primary Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-secondary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 6h12v12H6z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-secondary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 12c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-2-9c-4.97 0-9 4.03-9 9H0l4 4 4-4H5c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.51 0-2.91-.49-4.06-1.3l-1.42 1.44C8.04 20.3 9.94 21 12 21c4.97 0 9-4.03 9-9s-4.03-9-9-9z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`ControlPanel component should match snapshot for ongoing simulation 1`] = `
<div
  className="MuiPaper-root MuiCard-root controlPanel undefined MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-primary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-primary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 7.27l4.28 10.43-3.47-1.53-.81-.36-.81.36-3.47 1.53L12 7.27M12 2L4.5 20.29l.71.71L12 18l6.79 3 .71-.71L12 2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-primary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-secondary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 6h12v12H6z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-secondary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 12c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-2-9c-4.97 0-9 4.03-9 9H0l4 4 4-4H5c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.51 0-2.91-.49-4.06-1.3l-1.42 1.44C8.04 20.3 9.94 21 12 21c4.97 0 9-4.03 9-9s-4.03-9-9-9z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`ControlPanel component should match snapshot for stopped simulation 1`] = `
<div
  className="MuiPaper-root MuiCard-root controlPanel undefined MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root Mui-disabled MuiFab-primary Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root Mui-disabled MuiFab-primary Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 7.27l4.28 10.43-3.47-1.53-.81-.36-.81.36-3.47 1.53L12 7.27M12 2L4.5 20.29l.71.71L12 18l6.79 3 .71-.71L12 2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
  >
    <button
      className="MuiButtonBase-root MuiFab-root Mui-disabled MuiFab-primary Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiFab-root Mui-disabled MuiFab-secondary Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 6h12v12H6z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="controlButton"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiFab-root MuiFab-secondary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiFab-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 12c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-2-9c-4.97 0-9 4.03-9 9H0l4 4 4-4H5c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.51 0-2.91-.49-4.06-1.3l-1.42 1.44C8.04 20.3 9.94 21 12 21c4.97 0 9-4.03 9-9s-4.03-9-9-9z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
